cpu_scheduling_algorithms/round_robin.cpp

```diff
@@ -0,0 +1,207 @@
+                ready.push(i);
+                i++;
+            } else {
+                break;
+            }
+        }
+
+        // Main Round Robin scheduling loop
+        while (!ready.empty()) {
+            int j = ready.front();
+            ready.pop();
+            temp = t;
+
+            // Execute a job for a time quantum or until completion
+            if (taskList[j].remainingBurstTime < tm) {
+                t += taskList[j].remainingBurstTime;
+            } else {
+                t += tm;
+            }
+
+            taskList[j].remainingBurstTime =
+                max(taskList[j].remainingBurstTime - tm,
+                    0);  // Update remaining burst time
+
+            // Add jobs that arrive while a job is executing and have remaining
+            // burst time
+            for (int i = 0; i < num; i++) {
+                if (taskList[i].arrivalTime <= t &&
+                    taskList[i].arrivalTime > temp &&
+                    taskList[i].remainingBurstTime > 0) {
+                    ready.push(i);
+                }
+            }
+
+            // If the job is completed, calculate turnaround time and waiting
+            // time
+            if (taskList[j].remainingBurstTime > 0) {
+                ready.push(j);
+            } else {
+                taskList[j].completionTime = t;
+                taskList[j].turnAroundTime =
+                    taskList[j].completionTime - taskList[j].arrivalTime;
+                taskList[j].waitingTime =
+                    taskList[j].turnAroundTime - taskList[j].burstTime;
+                totalWaitingTime += taskList[j].waitingTime;
+                totalTurnAroundTime += taskList[j].turnAroundTime;
+            }
+        }
+
+        // Calculate and display average waiting time and average turnaround
+        // time
+        avgWaitingTime =
+            static_cast<double>(totalWaitingTime) / static_cast<double>(num);
+        avgTurnAroundTime =
+            static_cast<double>(totalTurnAroundTime) / static_cast<double>(num);
+
+        return {avgWaitingTime, avgTurnAroundTime};
+    }
+
+    void display() {
+        cout << "Status of all the proceses post completion :" << endl;
+        cout << "id\tname\tarrivalTime\tburstTime\tcompletionTime\twaitingTim"
+                "e\tTurnAroun"
+                "dTime"
+             << endl;
+        for (const auto& i : taskList) {
+            cout << i.id << "\t" << i.name << "\t" << i.arrivalTime << "\t"
+                 << i.burstTime << "\t" << i.completionTime << "\t"
+                 << i.waitingTime << "\t" << i.turnAroundTime << endl;
+        }
+    }
+};
+/** @brief functoin for testing round robin algorithm
+ *  @param input number of jobs and each job's id name arrival time and burst
+ * time
+ * @returns void
+ */
+static void test() {
+    vector<Job> t;
+    Job j1(1, "P1", 0, 5);
+    Job j2(2, "P2", 1, 4);
+    Job j3(3, "P3", 2, 2);
+    Job j4(4, "P4", 4, 1);
+    t.push_back(j1);
```

Comment:
add test cases for negative numbers or use unsigned in your algorithm

Reply:
Converted time related variables to unsigned int datatype in new commit